BUG: special: Fix hyperu for b = 1, small x, small a

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #15650 
Ticks a box on https://github.com/scipy/scipy/issues/20223
#### What does this implement/fix?
<!--Please explain your changes.-->
This PR uses DLMF 13.3.7 as suggested by @mdhaber in https://github.com/scipy/scipy/issues/15650#issuecomment-2323741880 to improve accuracy of `hyperu` when `b == 1` and `x` and `a` are close to zero. On `main`, `hyperu` is returning garbage values for such input. I think the underlying issue is that Zhang and Jin weren't careful when implementing the hypergeometric functions, and sometimes used formula outside of the domains in which they are valid. @mdhaber proposed to use @mdhaber when `b == 1`, `|x| < 1` and `|a| < 1`, based on numerical experiments described in the linked comment above, and I agree with his findings. I've added test cases covering this region.

#### Additional information
<!--Any additional information you think is important.-->
`hyperu` involves wrappers on top of wrappers. There is a wrapper `hyperu` in `_hypergeometric.pxd` which calls a wrapper `hypU_wrap` in `xsf/specfun.h` which calls `chgu` in `xsf/specfun/specfun.h`. The lowest level is ilhan's direct translation of Zhang and Jin's Fortran code, the middle layer is based on the old `specfun_wrappers` which called functions from `specfun.f`, and the outer Cython wrapper has some fixes for special cases patched in. I chose to make @mdhaber's suggested fix in the outer Cython wrapper, which is the most convenient spot. 

Obviously, this wrappers on top of wrappers architecture is not ideal. The plan is to eventually rewrite `hyperu` along the lines of https://github.com/scipy/scipy/pull/20089, and lift acceptably accurate functions up out of `xsf/specfun/specfun.h` as was done for exponential integrals here: https://github.com/scipy/scipy/pull/21210/files#diff-5fbee505531722bdef130b290e08018cb6291bfb85a3a84810c155c28e496210.

steppi committed:
Use DLMF 13.3.7 to fix gh-15650

Co-authored-by: Matt Haberland <mdhaber@mit.edu>

steppi committed:
Add test cases for gh-15650

mdhaber commented on scipy/special/tests/test_hypergeometric.py:
```diff
@@ -23,6 +23,41 @@ def test_special_cases(self):
+    def test_gh_15650(self, a, b, x, expected):
+        # See https://github.com/scipy/scipy/issues/15650
+        #
+        # Reference values computed with mpmath
```
Verified:
```python3
import numpy as np
from mpmath import mp
mp.dps = 100
for row in table:
    a, b, x, ref = (float(z) for z in row)
    np.testing.assert_equal(float(mp.hyperu(a, b, x)), ref)
```

mdhaber commented on scipy/special/tests/test_hypergeometric.py:
```diff
@@ -23,6 +23,41 @@ def test_special_cases(self):
 
+    @pytest.mark.parametrize(
+        'a,b,x,expected',
+        [(1e-15, 1.0, 1e-15, 1.0000000000000346),
```
I think negative `a` should be included in the tests. We might find that there is still a rough edge as `a` approaches `-1` because it will call `hyperu` with `a + 1`. If this can recurse the call  to `hyperu`, I think it would work. 
(If that won't work, we might need to adjusting the criterion `abs(a) < 1` to something like `abs(a) < 0.9`, but it might leave a sliver of the problematic behavior unless we fine-tune the threshold), or we could use the recurrence relation in the other direction, which I'd need to test.)

steppi commented on scipy/special/_hypergeometric.pxd:
```diff
@@ -26,5 +26,9 @@ cdef inline double hyperu(double a, double b, double x) noexcept nogil:
         else:
             # DLMF 13.2.14-15 and 13.2.19-21
             return cephes_poch_wrap(1.0 - b + a, -a)
+    if b == 1 and abs(a) < 1 and abs(x) < 1:
```
Just an oversight on my part, you're right. It should just be `x`.

steppi commented on scipy/special/_hypergeometric.pxd:
```diff
@@ -26,5 +26,9 @@ cdef inline double hyperu(double a, double b, double x) noexcept nogil:
+    if b == 1 and abs(a) < 1 and abs(x) < 1:
+        # DLMF 13.3.7. Fixes gh-15650
+        a = a + 1
+        return -(b - 2*a - x)*hypU_wrap(a, b, x) - a*(a - b + 1)*hypU_wrap(a + 1, b, x)
```
I'm not sure how much impact it actually has on performance, but it saves us from repeating the checks for `nan`, `x <  0`, `x == 0` on the previous lines.

steppi replied:
I see, if  `a < 0` then the recurrence still kicks us back into the problematic region. Yes, it needs to be `hyperu`.

mdhaber replied:
Not really. We just had to pick the threshold more carefully. See below.

steppi committed:
Adjust region in hyperu where DLMF 13.3.7 is used

steppi committed:
Add many new test cases for hyperu

steppi commented:
Thanks @mdhaber. You're right, there are issues when `a` approaches `-1`. I've adjusted the condition for applying DLMF 13.3.7 to `b == 1 and abs(a) < 0.9 and x < 1` as you suggested above. I've added a boatload of new test cases, and everything seems to be fine.

steppi committed:
Appease the linter [lint only]

steppi commented:
I'll make a `[lint only]` fix to appease the linter if CI settles.

mdhaber commented:
<img width="582" alt="image" src="https://github.com/user-attachments/assets/edfdaff7-0ade-4cfe-86b9-f681d4d13b99">

Looking in log-space was nice to get a rough idea of what was going on, but it wasn't great for choosing the threshold.

In main, this seems to be the edge of the problem area, this time with linear scale on the `a` axis.

<img width="583" alt="image" src="https://github.com/user-attachments/assets/ece6c9b3-9b5d-4dcc-b23f-db6c19d33a2f">

So it looks like `abs(a) < 0.5` would be a better threshold since it would keep up well away from the problem area on both sides.

steppi committed:
Adjust region where DLMF 13.3.7 is applied

steppi committed:
Adjust how DLMF 13.3.7 is evaluated

steppi committed:
Update hyperu test strategy

steppi commented:
Thanks @mdhaber. I took a look at this again. It's a good idea to cast a wide net with the tests by just checking if the values are not excessively large. In the actual test file I check

```python
        a = np.linspace(-0.5, 0.5, 500)
        x = np.linspace(1e-6, 1e-1, 500)
        a, x = np.meshgrid(a, x)
        results = sc.hyperu(a, 1.0, x)
        assert np.all(np.abs(results) < 1e3)
```

so that the tests don't run for too long, but locally I did

```python
        a = np.linspace(-0.5, 0.5, 10000)
        x = np.logspace(-6, 0, 1000)
```

which is how I ended up selecting the boundary points for using DLMF 13.3.7. Based on comparison with `mpmath` I observed that  max relative error is better when you don't use DLMF 13.3.7 for `a` where the problematic behavior is not seen, so I tried to make the interval as tight as I thought was reasonable.

I ran the script found in the details section below which looks at a range of cases with `|a| < 1` for various x, and for each `a` finds the maximum relative error over each `x`. Here's a plot of the log10 of these max relative errors as a function of  `a` both when and when not applying DLMF 13.3.7. (This is with recursively calling to `hyperu` inside the Cython code instead of calling `hypU_wrap` so that the recurrence won't leave us back in the problematic region. This is what I was alluding to in my comment above. Note in the plot that for `a` close to 1 it's still better to make the adjustment. At the time I wrote the comment, I was considering using the recurrence when `a` is close to 1, but I don't think such complexity is warranted now.)

![hyperu_example](https://github.com/user-attachments/assets/f3a68ecc-4d39-4e5a-a513-6d5c4cac6c97)

My assumption is that outside of `a` close to -1, as long as whatever catastrophic failure is happening to cause gh-15650 doesn't occur, it's better not to use the recurrence. So I did the check mentioned above  that doesn't involve `mpmath`, found the largest and smallest values of `a` where the failure occurs, and set the boundary a little wider.


<details>

```python

import itertools as it
import numpy as np
import pandas as pd
import scipy.special as special
        
from mpmath import mp
        
cases = []
for a, x in it.product(
        np.concat(
            [np.logspace(-15, -1e-6, 100), np.linspace(0.1, 0.99, 100)]
        ),
        repeat=2,
):
    with mp.workdps(100):
        cases.append((float(a), 1.0, float(x), float(mp.hyperu(a, 1.0, x)), special.hyperu(a, 1.0, x)))
        cases.append((-float(a), 1.0, float(x), float(mp.hyperu(-a, 1.0, x)), special.hyperu(-a, 1.0, x)))

df = pd.DataFrame(cases, columns=["a", "b", "x", "expected", "observed"])
df["relative_error"] = abs(df.expected - df.observed) / abs(df.expected)
df["log10_re"] = df.relative_error.apply(lambda x: -16 if x == 0 else np.log10(x))
max_re_vs_a = df.groupby("a", as_index=False).log10_re.max()
```

</details>

oops. I angered the linter again.

steppi committed:
Appease the linter [lint only]

mdhaber commented on scipy/special/tests/test_hypergeometric.py:
```diff
@@ -25,39 +25,66 @@ def test_nan_inputs(self, a, b, x):
+        # was impacted by gh-15650 by making sure there are no excessively large
+        # results, as were reported there.
+        a = np.linspace(-0.5, 0.5, 500)
+        x = np.linspace(1e-6, 1e-1, 500)
```
Just checking because it looks funny - is this supposed to be logspace? (But based on having worked with this failure a lot, I would agree that linear spacing is probably better here if `a` is linspaced.)

steppi replied:
Yeah, I intended this to be `linspace`.

mdhaber reviewed (commented):
Ok, I just need one more look from a computer.

mdhaber commented:
Thanks @steppi!
